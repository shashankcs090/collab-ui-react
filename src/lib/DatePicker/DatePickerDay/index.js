import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@collab-ui/react';
import {
  getDate,
  getMonth,
  isDayDisabled,
  isSameDay,
  now,
} from '@collab-ui/react/DatePicker/date_utils';
import moment from 'moment';

class DatePickerDay extends React.Component {
  static displayName = 'DatePickerDay';

  handleClick = e => {
    const { handleDayClick } = this.context;
    const { day } = this.props;
    return (
      handleDayClick
      && handleDayClick(e, day)
    );
  };

  render() {
    const { day, month } = this.props;
    const { selected, focus } = this.context;

    const isOutsideMonth = month !== getMonth(day);
    const isSelected = isSameDay(day, selected);
    const isToday = isSameDay(day, now());
    const disabled = isDayDisabled(day, this.props);
    const hasFocus = isSameDay(day, focus);

    return (
      <Button
        circle
        size={28}
        disabled={disabled}
        className={
          'cui-datepicker__day' +
          `${(isSelected && ` cui-datepicker__day--selected`) || ''}` +
          `${(hasFocus && ` cui-datepicker__day--focus`) || ''}` +
          `${(isToday && ` cui-datepicker__day--today`) || ''}` +
          `${(isOutsideMonth && ` cui-datepicker__day--outside-month`) || ''}`
        }
        onClick={this.handleClick}
        ariaLabel={`${getDate(day)}`}
        aria-selected={isSelected}
        tabIndex={-1}
      >
        {getDate(day)}
      </Button>
    );
  }
}

DatePickerDay.contextTypes = {
  focus: PropTypes.instanceOf(moment),
  handleDayClick: PropTypes.func,
  selected: PropTypes.instanceOf(moment),
};

DatePickerDay.propTypes = {
  day: PropTypes.instanceOf(moment).isRequired,
  month: PropTypes.number.isRequired,
};

export default DatePickerDay;
