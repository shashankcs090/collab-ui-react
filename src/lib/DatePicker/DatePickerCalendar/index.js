import React from 'react';
import PropTypes from 'prop-types';
import DatePickerMonth from '@collab-ui/react/DatePicker/DatePickerMonth';
import { Icon }  from '@collab-ui/react';
import {
  addDays,
  addMonths,
  getLocaleData,
  getStartOfWeek,
  getWeekdayMinInLocale,
  isSameDay,
  localizeDate,
  now,
  shouldNextMonthDisable,
  shouldPrevMonthDisable,
  subtractMonths,
} from '@collab-ui/react/DatePicker/date_utils';
import moment from 'moment';

class DatePickerCalendar extends React.Component {
  static displayName = 'DatePickerCalendar';

  state = {
    date: null,
  };

  componentDidMount () {
    const { focus, selected } = this.context;
    this.setDate(focus || selected || now());
  }

  componentDidUpdate (prevProps, prevState, prevContext) {
    const { focus } = this.context;
    if (
      focus &&
      !isSameDay(prevContext.focus, focus)
    ) {
      this.setDate(focus);
    }
  }

  setDate = (date, cb) => {
    this.setState({
      date
    }, cb);
  };

  increaseMonth = event => {
    const { handleMonthChange } = this.context;
    const { date } = this.state;
    this.setDate(
      addMonths(date.clone(), 1),
      () => handleMonthChange && handleMonthChange(event, this.state.date)
    );
  };

  decreaseMonth = event => {
    const { handleMonthChange } = this.context;
    const { date } = this.state;
    this.setDate(
      subtractMonths(date.clone(), 1),
      () => handleMonthChange && handleMonthChange(event, this.state.date)
    );
  };

  render() {

    const { date } = this.state;
    const {
      locale,
      monthFormat,
      nextArialLabel,
      previousArialLabel,
      ...otherProps
    } = this.props;

    const renderMonthName = () => {
      return (
        <div className='cui-datepicker__navigation--current-month'>
          {localizeDate(date, locale).format(monthFormat)}
        </div>
      );
    };

    const renderPreviousMonthButton = () => {
      const { minDate } = this.props;
      const allPrevDaysDisabled = shouldPrevMonthDisable(date, minDate);
      return (
        <Icon
          ariaLabel={previousArialLabel}
          isClickable
          disabled={allPrevDaysDisabled}
          onClick={this.decreaseMonth}
          name='arrow-left_16'
        />
      );
    };

    const renderNextMonthButton = () => {
      const { maxDate } = this.props;
      const allNextDaysDisabled = shouldNextMonthDisable(date, maxDate);

      return (
        <Icon
          ariaLabel={nextArialLabel}
          isClickable
          disabled={allNextDaysDisabled}
          onClick={this.increaseMonth}
          name='arrow-right_16'
        />
      );
    };

    const header = () => {
      const startOfWeek = getStartOfWeek(date.clone());
      const dayNames = [];
      return dayNames.concat(
        [0, 1, 2, 3, 4, 5, 6].map(offset => {
          const day = addDays(localizeDate(startOfWeek, locale), offset);
          const localeData = getLocaleData(day);
          const weekDayName = getWeekdayMinInLocale(localeData, day);
          return (
            <div key={offset} className='cui-datepicker__day--name'>
              {weekDayName}
            </div>
          );
        })
      );
    };

    const renderMonth = () => {
      return (
        <div className='cui-datepicker__month-container'>
          <div className='cui-datepicker__header'>
            <div className='cui-datepicker__navigation'>
              {renderMonthName()}
              <div className='cui-datepicker__navigation--buttons'>
                {renderPreviousMonthButton()}
                {renderNextMonthButton()}
              </div>
            </div>
            <div className='cui-datepicker__day--names'>
              {header()}
            </div>
          </div>
          <DatePickerMonth
            day={date}
            {...otherProps}
          />
        </div>
      );
    };

    return (
      <div>
        {date && renderMonth()}
      </div>
    );
  }
}

DatePickerCalendar.contextTypes = {
  focus: PropTypes.instanceOf(moment),
  handleMonthChange: PropTypes.func,
  selected: PropTypes.instanceOf(moment),
};

DatePickerCalendar.propTypes = {
  locale: PropTypes.string,
  maxDate: PropTypes.instanceOf(Date),
  minDate: PropTypes.instanceOf(Date),
  monthFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  nextArialLabel: PropTypes.string,
  previousArialLabel: PropTypes.string,
};

DatePickerCalendar.defaultProps = {
  locale: 'en',
  maxDate: null,
  minDate: null,
  monthFormat: 'MMMM YYYY',
  nextArialLabel: 'next',
  previousArialLabel: 'previous',
};

export default DatePickerCalendar;
