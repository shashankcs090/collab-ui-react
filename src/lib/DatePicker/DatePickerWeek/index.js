import React from 'react';
import PropTypes from 'prop-types';
import DatePickerDay from '@collab-ui/react/DatePicker/DatePickerDay';
import { addDays, getStartOfWeek } from '@collab-ui/react/DatePicker/date_utils';
import moment from 'moment';

class DatePickerWeek extends React.PureComponent {
  static displayName = 'DatePickerWeek';

  render() {
    const { day, ...otherProps } = this.props;

    const renderDays = () => {
      const startOfWeek = getStartOfWeek(day.clone());

      const days = [0, 1, 2, 3, 4, 5, 6].map(offset => {
        const day = addDays(startOfWeek.clone(), offset);
        return (
          <DatePickerDay
            key={offset}
            day={day}
            {...otherProps}
          />
        );
      });

      return days;
    };

    return (
      <div className="cui-datepicker__week">
        {renderDays()}
      </div>
    );
  }
}

DatePickerWeek.propTypes = {
  day: PropTypes.instanceOf(moment).isRequired,
};

export default DatePickerWeek;
