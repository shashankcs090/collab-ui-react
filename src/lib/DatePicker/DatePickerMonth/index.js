import React from 'react';
import PropTypes from 'prop-types';
import DatePickerWeek from '@collab-ui/react/DatePicker/DatePickerWeek';
import {
  addWeeks,
  getMonth,
  getStartOfMonth,
  getStartOfWeek,
  isSameMonth,
} from '@collab-ui/react/DatePicker/date_utils';
import moment from 'moment';

class DatePickerMonth extends React.Component {
  static displayName = 'DatePickerMonth';

  renderWeeks = () => {
    const { day, ...otherProps } = this.props;

    let i = 0;
    let currentWeekStart = getStartOfWeek(
      getStartOfMonth(day.clone())
    );

    const weeks = [];
    const month = getMonth(day);

    do {
      weeks.push(
        <DatePickerWeek
          key={i++}
          day={currentWeekStart}
          month={month}
          {...otherProps}
        />
      );
      currentWeekStart = addWeeks(currentWeekStart.clone(), 1);

    } while(isSameMonth(currentWeekStart, day));

    return weeks;
  };


  render() {
    return (
      <div className='cui-datepicker__month'>
        {this.renderWeeks()}
      </div>
    );
  }
}

DatePickerMonth.propTypes = {
  day: PropTypes.instanceOf(moment).isRequired,
};

export default DatePickerMonth;
