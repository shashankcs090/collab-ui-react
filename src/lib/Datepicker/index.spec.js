import React from 'react';
import { mount, shallow } from 'enzyme';
import DatePicker from '../Datepicker';
import moment from 'moment-timezone';

describe('tests for <DatePicker />', () => {
  const TEST_DATE = '2018-04-01';
  let day;

  beforeAll(() => {
    // setting timezone to CST
    moment.tz.setDefault('Asia/Kolkata');
    day = moment(TEST_DATE);
  });

  afterAll(() => {
    moment.tz.setDefault();
  });

  it("should show the calendar when focusing on the date input", () => {
    const container = shallow(
      <DatePicker initialSelection={day}/>
    );
    expect(container).toMatchSnapshot();
  });

  it("should focus the initialSelection day if passed", () => {
    const container = mount(
      <DatePicker initialSelection={day}/>
    );
    container.find('input').simulate('mousedown');
    container.update();
    expect(container.find('.cui-button.cui-datepicker__day--focus').text()).toEqual("1");
  });

  it("on press of enter should select the date", () => {
    const container = mount(
      <DatePicker initialSelection={day} shouldCloseOnSelect={false}/>
    );
    container.find('input').simulate('mousedown');
    container.update();
    container.find('.cui-button.cui-datepicker__day--focus').simulate('click');
    expect(container.find('.cui-button.cui-datepicker__day--selected').text()).toEqual("1");
    expect(container.find('input').props().value).toEqual('04/01/2018');
  });

  it("when shouldCloseOnSelect is true should close the DatePicker", () => {
    const container = mount(
        <DatePicker initialSelection={day} shouldCloseOnSelect={true}/>
    );
    container.find('input').simulate('mousedown');
    container.update();
    container.find('.cui-button.cui-datepicker__day--focus').simulate('click');
    expect(container.find('.cui-event-overlay__children').length).toEqual(0);
    expect(container.find('input').props().value).toEqual('04/01/2018');
  });

  it("should handle keyBoard keys", () => {
    const container = mount(
        <DatePicker initialSelection={day} shouldCloseOnSelect={false}/>
    );
    container.find('input').simulate('mousedown');
    container.update();
    // right
    container.find('input').simulate('keydown', { which: 39 });
    expect(container.find('.cui-button.cui-datepicker__day--focus').text()).toEqual("2");
    // down
    container.find('input').simulate('keydown', { which: 40 });
    expect(container.find('.cui-button.cui-datepicker__day--focus').text()).toEqual("9");
    // up
    container.find('input').simulate('keydown', { which: 38 });
    expect(container.find('.cui-button.cui-datepicker__day--focus').text()).toEqual("2");
    // left
    container.find('input').simulate('keydown', { which: 37 });
    expect(container.find('.cui-button.cui-datepicker__day--focus').text()).toEqual("1");

    container.find('input').simulate('keydown', { which: 13 });
    expect(container.find('.cui-button.cui-datepicker__day--selected').text()).toEqual("1");
    expect(container.find('input').props().value).toEqual('04/01/2018');
  });
});
