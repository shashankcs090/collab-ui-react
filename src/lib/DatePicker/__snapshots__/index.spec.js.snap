// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <DatePicker /> should show the calendar when focusing on the date input 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DatePicker
    className=""
    direction="bottom-center"
    filterDate={null}
    isDynamic={true}
    locale="en"
    maxDate={null}
    minDate={null}
    monthFormat="MMMM YYYY"
    nextArialLabel="next"
    onChange={null}
    onMonthChange={null}
    onSelect={null}
    previousArialLabel="previous"
    selectedDate={2018-03-31T18:30:00.000Z}
    shouldCloseOnSelect={true}
    showArrow={false}
  >
    <div
      className="trigger"
    >
      Select Date
    </div>
  </DatePicker>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="trigger"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Select Date
        </div>,
        <EventOverlay
          allowClickAway={true}
          anchorNode={null}
          checkOverflow={true}
          className=""
          close={[Function]}
          direction="bottom-center"
          isDynamic={true}
          isOpen={false}
          maxHeight={null}
          maxWidth={null}
          showArrow={false}
          style={null}
          targetOffset={
            Object {
              "horizontal": 0,
              "vertical": 0,
            }
          }
        >
          <DatePickerCalendar
            filterDate={null}
            focus={"2018-03-31T18:30:00.000Z"}
            locale="en"
            maxDate={null}
            minDate={null}
            monthFormat="MMMM YYYY"
            nextArialLabel="next"
            previousArialLabel="previous"
            selected={"2018-03-31T18:30:00.000Z"}
            selectedDate={2018-03-31T18:30:00.000Z}
          />
        </EventOverlay>,
      ],
      "className": "cui-datepicker-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Select Date",
          "className": "trigger",
          "onClick": [Function],
          "onKeyDown": [Function],
        },
        "ref": [Function],
        "rendered": "Select Date",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowClickAway": true,
          "anchorNode": null,
          "checkOverflow": true,
          "children": <DatePickerCalendar
            filterDate={null}
            focus={"2018-03-31T18:30:00.000Z"}
            locale="en"
            maxDate={null}
            minDate={null}
            monthFormat="MMMM YYYY"
            nextArialLabel="next"
            previousArialLabel="previous"
            selected={"2018-03-31T18:30:00.000Z"}
            selectedDate={2018-03-31T18:30:00.000Z}
          />,
          "className": "",
          "close": [Function],
          "direction": "bottom-center",
          "isDynamic": true,
          "isOpen": false,
          "maxHeight": null,
          "maxWidth": null,
          "showArrow": false,
          "style": null,
          "targetOffset": Object {
            "horizontal": 0,
            "vertical": 0,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "filterDate": null,
            "focus": "2018-03-31T18:30:00.000Z",
            "locale": "en",
            "maxDate": null,
            "minDate": null,
            "monthFormat": "MMMM YYYY",
            "nextArialLabel": "next",
            "previousArialLabel": "previous",
            "selected": "2018-03-31T18:30:00.000Z",
            "selectedDate": 2018-03-31T18:30:00.000Z,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="trigger"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Select Date
          </div>,
          <EventOverlay
            allowClickAway={true}
            anchorNode={null}
            checkOverflow={true}
            className=""
            close={[Function]}
            direction="bottom-center"
            isDynamic={true}
            isOpen={false}
            maxHeight={null}
            maxWidth={null}
            showArrow={false}
            style={null}
            targetOffset={
              Object {
                "horizontal": 0,
                "vertical": 0,
              }
            }
          >
            <DatePickerCalendar
              filterDate={null}
              focus={"2018-03-31T18:30:00.000Z"}
              locale="en"
              maxDate={null}
              minDate={null}
              monthFormat="MMMM YYYY"
              nextArialLabel="next"
              previousArialLabel="previous"
              selected={"2018-03-31T18:30:00.000Z"}
              selectedDate={2018-03-31T18:30:00.000Z}
            />
          </EventOverlay>,
        ],
        "className": "cui-datepicker-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Select Date",
            "className": "trigger",
            "onClick": [Function],
            "onKeyDown": [Function],
          },
          "ref": [Function],
          "rendered": "Select Date",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowClickAway": true,
            "anchorNode": null,
            "checkOverflow": true,
            "children": <DatePickerCalendar
              filterDate={null}
              focus={"2018-03-31T18:30:00.000Z"}
              locale="en"
              maxDate={null}
              minDate={null}
              monthFormat="MMMM YYYY"
              nextArialLabel="next"
              previousArialLabel="previous"
              selected={"2018-03-31T18:30:00.000Z"}
              selectedDate={2018-03-31T18:30:00.000Z}
            />,
            "className": "",
            "close": [Function],
            "direction": "bottom-center",
            "isDynamic": true,
            "isOpen": false,
            "maxHeight": null,
            "maxWidth": null,
            "showArrow": false,
            "style": null,
            "targetOffset": Object {
              "horizontal": 0,
              "vertical": 0,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "filterDate": null,
              "focus": "2018-03-31T18:30:00.000Z",
              "locale": "en",
              "maxDate": null,
              "minDate": null,
              "monthFormat": "MMMM YYYY",
              "nextArialLabel": "next",
              "previousArialLabel": "previous",
              "selected": "2018-03-31T18:30:00.000Z",
              "selectedDate": 2018-03-31T18:30:00.000Z,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
